workflow-manager/main.go

```diff
@@ -420,16 +420,27 @@ func (b *bucket) listFilesGS(ctx context.Context) ([]string, error) {
 	log.Printf("looking for ready batches in gs://%s as (ambient service account)", b.bucketName)
 	var output []string
 	it := bkt.Objects(ctx, query)
+
+	// https://cloud.google.com/storage/docs/json_api/v1/objects/list
+	// We can have up to: 1000
+	p := iterator.NewPager(it, 100, "")
+
 	for {
-		attrs, err := it.Next()
-		if err == iterator.Done {
-			break
-		}
+		var objects []*storage.ObjectAttrs
+		nextPageToken, err := p.NextPage(&objects)
 		if err != nil {
-			return nil, fmt.Errorf("iterating on inputBucket objects from %q: %w", b.bucketName, err)
+			return nil, fmt.Errorf("storage.nextPage: %w", err)
+		}
+
+		for _, obj := range objects {
+			output = append(output, obj.Name)
+		}
```

Comment:
[`Pager.NextPage`](https://godoc.org/google.golang.org/api/iterator#Pager.NextPage) is documented to append the next page's contents to the slice passed in. So I think you could save yourself this inner `for` loop and all the copying it does if you instead declare `var objects` outside the outer loop, and pass `objects` into `p.NextPage` until `nextPageToken == ""`.

Reply:
That would work, the only concern I have is that the code needs to keep a LOT more objects in memory, but they're all small metadata anyway...

Reply:
The iterator is a lazy iterator, so it can get rid of the objects it creates and just keep the names around, but I think that's just premature optimization at that point.